naga/src/front/wgsl/lower/mod.rs

```diff
@@ -2483,12 +2484,70 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
                         "select" => {
                             let mut args = ctx.prepare_args(arguments, 3, span);
 
-                            let reject = self.expression(args.next()?, ctx)?;
-                            let accept = self.expression(args.next()?, ctx)?;
+                            let reject_orig = args.next()?;
+                            let accept_orig = args.next()?;
+                            let mut values = [
+                                self.expression_for_abstract(reject_orig, ctx)?,
+                                self.expression_for_abstract(accept_orig, ctx)?,
+                            ];
                             let condition = self.expression(args.next()?, ctx)?;
 
                             args.finish()?;
 
+                            fn expr_ty<'a>(
+                                ctx: &'a ExpressionContext<'_, '_, '_>,
+                                expr: Handle<ir::Expression>,
+                            ) -> &'a ir::TypeInner {
+                                ctx.typifier()[expr].inner_with(&ctx.module.types)
+                            }
+                            let diag_deets =
+                                |module: &ir::Module, ty: &ir::TypeInner, orig_expr| {
+                                    (
+                                        ctx.ast_expressions.get_span(orig_expr),
+                                        format!("`{:?}`", ty.for_debug(&module.types)),
```

Comment:
Ideally we should hold on to the `TypeResolution` for these expressions, and use `ctx.as_diagnostic_display(resolution)` for generating diagnostics, as by the time we've reduced that to a `TypeInner`, we've lost the names of struct types.